README.md
MetaGxBreast
========

ExperimentalHubData package to be submitted containing 39 breast cancer datasets


Installation
========

The package can be installed by running the below two lines

if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")

BiocManager::install("MetaGxBreast")


inst/NEWS
CHANGES IN VERSION 0.99.0
-------------------------

NEW FEATURES

    o This is the first iteration of the MetaGxBreast package. The package contains 39 breast cancer Esets
    
CHANGES IN VERSION 0.99.5
-------------------------

NEW FEATURES

    o loadBreastDatasets() added to load in summarized experiment datasets as oppose to the outdated esets
    o Fixed incorrect time to distant metastases present in GSE58644


tests/testthat/test_loadBreastEsets.R

library(MetaGxBreast)

context("Checking loadBreastEsets")


test_that("ensure datasets and duplicates are properly loaded from the hub and package", {
  esetsAndDuplicates = MetaGxBreast::loadBreastEsets(loadString = c("CAL", "DFHCC", "DFHCC2", "DFHCC3", "DUKE", "DUKE2", "EMC2"))
  esets = esetsAndDuplicates$esets
  expect_equal(class(esets[[1]])[1], "ExpressionSet")
  expect_equal(class(esets[[1]]@assayData$exprs), "matrix")

})


tests/testthat/test_loadBreastDatasets.R
library(MetaGxBreast)

context("Checking loadBreastEsets")


test_that("ensure datasets and duplicates are properly loaded from the hub and package", {
  esetsAndDuplicates = MetaGxBreast::loadBreastEsets()
  esets = esetsAndDuplicates$esets
  #still class warning regardless of replacing with is
  expect_equal(is(esets[[1]])[1], "ExpressionSet")
  expect_equal(is(esets[[1]]@assayData$exprs)[1], "matrix")
})


R/loadBreastDatasets.R
#' Function to load breast cancer SummarizedExperiment objectsfrom the Experiment Hub
#'
#' This function returns breast cancer datasets from the hub and a vector of patients from the datasets that are duplicates based on a spearman correlation > 0.98
#' @param rescale apply centering and scaling to the expression sets (default FALSE)
#' @param minNumberGenes an integer specifying to remove expression sets with less genes than this number (default 0)
#' @param minNumberEvents an integer specifying how man survival events must be in the dataset to keep the dataset (default 0)
#' @param minSampleSize an integer specifying the minimum number of patients required in a summarizedExperiment (default 0)
#' @param keepCommonOnly remove entrezIDs not common to all datasets (default FALSE)
#' @param imputeMissing remove patients from datasets with missing expression values
#' @param removeDuplicates remove patients with a Spearman correlation greater than or equal to 0.98 with other patient expression profiles (default TRUE)
#' @return a list with 2 elements. The First element named summarizedExperiments contains the datasets. The second element named duplicates contains
#' a vector with patient IDs for the duplicate patients (those with  Spearman correlation greater than or equal to 0.98 with other patient expression profiles).
#' @export
#' @importFrom Biobase esApply featureNames sampleNames exprs pData experimentData ExpressionSet
#' @importFrom lattice levelplot
#' @importFrom impute impute.knn
#' @importFrom ExperimentHub ExperimentHub
#' @importFrom AnnotationHub query
#' @importFrom stats complete.cases sd quantile
#' @importFrom SummarizedExperiment SummarizedExperiment assays assayNames colData rowData
#' @examples
#'
#' experimentsAndDups = loadBreastDatasets()


loadBreastDatasets = function(rescale = FALSE, minNumberGenes = 0,
                                minNumberEvents = 0, minSampleSize = 0,
                                keepCommonOnly = FALSE, imputeMissing = FALSE, removeDuplicates = FALSE)
{
  duplicates = NULL
  #if(getRversion() >= "2.15.1")  utils::globalVariables(c("."), add = F)
  ## -----------------------------------------------------------------------------
  ## needed functions
  ## -----------------------------------------------------------------------------

  ##recursive intersect function
  intersectMany <- function(lst){
    ## Find the intersection of multiple vectors stored as elements of a
    ## list, through a tail-recursive function.
    if (length(lst)==2){
      return(intersect(lst[[1]],lst[[2]]))
    }else{
      return(intersectMany(c(list(intersect(lst[[1]],lst[[2]])),lst[seq(-1, -2)])))
    }
  }


  ## -----------------------------------------------------------------------------
  ##load the summarizedExperiments
  ## -----------------------------------------------------------------------------

  hub = ExperimentHub::ExperimentHub()
  #AnnotationHub::possibleDates(hub)
  #query(eh, c("MetaGxOvarian", "SummarizedExperiment"))
  breastData = query(hub, c("MetaGxBreast", "SummarizedExperiment"))
  #pancreas issues: loading dataset 6, but missing /v1/ for 3 datasets
  dataList <- list()
  for(i in seq_len(length(breastData)))
  {
    dataList[[i]] = breastData[[names(breastData)[i]]]
    names(dataList)[i] = breastData[i]$title
  }
  names(dataList) = gsub("_sumexp", "", names(dataList), ignore.case = TRUE)

  ## -----------------------------------------------------------------------------
  ##Explicit removal of samples from specified datasets:
  ## -----------------------------------------------------------------------------
  delim <- ":"   ##This is the delimiter used to specify dataset:sample,

  ## same as used in metagx getbrcadata
  #load("inst\\extdata\\BenDuplicate.rda")
  #source(system.file("extdata", "patientselection.config", package="MetaGxOvarian"))
  load(system.file("extdata", "duplicates.rda", package="MetaGxBreast"))

  rmix <- duplicates
  ii <- 1
  while (length(rmix) > ii){
    rmix <- rmix [!is.element(names(rmix), rmix[[ii]])]
    ii <- ii+1
  }
  rmix <- unique(unlist(rmix))
  rmix = substr(rmix, unlist(lapply(gregexpr("\\.", rmix), function(x) x[[1]][1]+1)), nchar(rmix))

  message("Clean up the summarizedExperiments")
  remInds = c()
  for (i in seq_len(length(dataList))){
    data <- dataList[[i]]
    include = TRUE
    ##rescale to z-scores
    if(rescale == TRUE){
      SummarizedExperiment::assay(data) = t(scale(t(SummarizedExperiment::assay(data))))
    }

    if(removeDuplicates == TRUE){
      keepix <- setdiff(colnames(SummarizedExperiment::assay(data)), rmix)
      if(length(keepix) != length(colnames(SummarizedExperiment::assay(data))))
      {
        keepix = which(!colnames(SummarizedExperiment::assay(data)) %in% rmix)
        data = data[ ,keepix]
      }
    }

    ##include study if it has enough samples and events:

    phenoData = colData(data)
    remData = TRUE
    if(nrow(phenoData) - sum(is.na(phenoData$vital_status)) >= minNumberEvents) remData = FALSE
    if(nrow(phenoData) - sum(is.na(phenoData$recurrence_status)) >= minNumberEvents) remData = FALSE
    if(nrow(phenoData) >= minSampleSize) remData = FALSE

    if(remData == TRUE){
      message(paste("excluding", names(dataList)[i], "(minNumberEvents or minSampleSize)"))
      remInds = c(remInds, i)
      include = FALSE
    }


    if(nrow(data) < minNumberGenes) {
      message(paste("excluding experiment hub dataset",names(dataList)[i],"(minNumberGenes)"))
      remInds = c(remInds, i)
      include = FALSE

    }

    if(imputeMissing == TRUE){
      notNaInds = which(colSums(is.na(SummarizedExperiment::assay(data))) == 0)
      data = data[, notNaInds]
      if(length(notNaInds) == 0){
        message(paste("excluding experiment hub dataset",names(dataList)[i],
                      "as every patient has at least 1 NA expression value (imputmissing = TRUE)"))
        remInds = c(remInds, i)
        include = FALSE

      }
    }
    if(include == TRUE) message(paste("including experiment hub dataset",names(dataList)[i]))
    ##    featureNames(eset) <- make.names(featureNames(eset))  ##should not do this, it is irreversible.
    dataList[[i]] <- data
    rm(data)
  }
  if(length(remInds) > 0) dataList[unique(remInds)] = NULL


  ##optionally take the intersection of genes common to all platforms:
  if(keepCommonOnly & length(dataList) > 0){
    features.per.dataset <- lapply(dataList, function(x) rowData(x)$EntrezGene.ID)
    intersect.genes <- intersectMany(features.per.dataset)
    dataList <- lapply(dataList, function(data){
      data <- data[which(rowData(data)$EntrezGene.ID %in% intersect.genes), ]
      return(data)
    })
  }
  if(length(dataList) == 0) warning("input values resulted in no datasets being returned")


  retList = list(dataList, duplicates)
  names(retList) = c("summarizedExperiments", "duplicates")
  return(retList)
}


inst/scripts/make-metadata.R
## validate with `AnnotationHub::readMetadataFromCsv("TENxBrainData")`
## (above pkg directory)

infoString = c("CAL", "DFHCC", "DFHCC2", "DFHCC3", "DUKE", "DUKE2", "EMC2", "EORTC10994", "EXPO",
               "FNCLCC", "GSE25066", "GSE32646","GSE48091","GSE58644","HLP","IRB", "KOO","LUND",
               "LUND2","MAINZ","MAQC2","MCCC","MDA4","METABRIC","MSK","MUG","NCCS","NCI","NKI",
               "PNC","STK","STNO2","TCGA","TRANSBIG","UCSF","UNC4","UNT","UPP","VDX")


main.data <- data.frame(
  Title = c("CAL", "DFHCC", "DFHCC2", "DFHCC3", "DUKE", "DUKE2", "EMC2", "EORTC10994", "EXPO",
            "FNCLCC", "GSE25066", "GSE32646","GSE48091","GSE58644","HLP","IRB", "KOO","LUND",
            "LUND2","MAINZ","MAQC2","MCCC","MDA4","METABRIC","MSK","MUG","NCCS","NCI","NKI",
            "PNC","STK","STNO2","TCGA","TRANSBIG","UCSF","UNC4","UNT","UPP","VDX"
  ),
  Description = paste(
    "",
    c(
      "Dataset of breast cancer patients from the University of California, San Francisco and the California Pacific Medical Center (United States)",
      "Dataset of breast cancer patients from Dana-Farber Harvard Cancer Center (United States)",
      "Dataset of breast cancer patients from Dana-Farber Harvard Cancer Center (United States)",
      "Dataset of breast cancer patients from Dana-Farber Harvard Cancer Center (United States)",
      "Dataset of breast cancer patients from Duke university hospital (United States)",
      "Dataset of breast cancer patients from Duke university hospital (United States)",
      "Dataset of breast cancer patients from Erasmus Medical Center (The Netherlands)",
      "Dataset of breast cancer patients from Trial number 10994 from the European Organization for Research and Treatment of Cancer Breast Cancer (Europe)",
      "Dataset of breast cancer patients from Expression project for oncology, large dataset of microarray data published by the International Genomics Consortium",
      "Dataset of breast cancer patients from F_d_ration Nationale des Centres de Lutte contre le Cancer (France)",
      "Dataset of breast cancer patients from Nuvera Biosciences",
      "Dataset of breast cancer patients from Osaka University",
      "Dataset of breast cancer patients from Karolinska Institutet",
      "Dataset of breast cancer patients from McGill University",
      "Dataset of breast cancer patients from University Hospital La Paz (Spain)",
      "Dataset of breast cancer patients from Dana Farber Cancer Institute",
      "Dataset of breast cancer patients from Koo Foundation Sun Yat-Sen Cancer Centre (Taiwan)",
      "Dataset of breast cancer patients from Lund University Hospital (Sweden)",
      "Dataset of breast cancer patients from Lund University Hospital (Sweden)",
      "Dataset of breast cancer patients from Mainz hospital (Germany)",
      "Dataset of breast cancer patients from Microarray quality control consortium (United States)",
      "Dataset of breast cancer patients from Peter MacCallum Cancer Centre (Australia)",
      "Dataset of breast cancer patients from MD Anderson Cancer Center (United States)",
      "2,000 clinically annotated primary fresh-frozen breast cancer specimens from tumour banks in the UK and Canada",
      "Dataset of breast cancer patients from Memorial Sloan-Kettering (United States)",
      "Dataset of breast cancer patients from Medical University of Graz (Austria)Medical University of Graz (Austria)",
      "Dataset of breast cancer patients from National Cancer Centre of Singapore (Singapore)",
      "Dataset of breast cancer patients from National Cancer Institute (United States)",
      "Dataset of breast cancer patients from National Kanker Instituut (The Netherlands)",
      "This dataset (GSE20711) Subseries of GSE20713 used to study epigenetic variations link to gene expression",
      "Dataset of breast cancer patients from Stockholm",
      "Dataset of breast cancer patients from Stanford/Norway (United States and Norway)",
      "Dataset of breast cancer patients from The Cancer Genome Atlas",
      "Dataset collected by the TransBIG consortium (Europe)",
      "Dataset of breast cancer patients from University of California, San Francisco (United States)",
      "Dataset of breast cancer patients from University of Northern California (United States)",
      "Cohort of untreated breast cancer patients from the Oxford Radcliffe (United Kingdom) and Karolinska (Sweden) hospital",
      "Dataset of breast cancer patients from Uppsala hospital (Sweden)",
      "Dataset of breast cancer patients from Veridex (The Netherlands)"
    )
  ),
  RDataPath = c(
    "MetaGxBreast/CAL.rda",          "MetaGxBreast/DFHCC.rda",      "MetaGxBreast/DFHCC2.rda",
    "MetaGxBreast/DFHCC3.rda",      "MetaGxBreast/DUKE.rda",      "MetaGxBreast/DUKE2.rda",
    "MetaGxBreast/EMC2.rda",     "MetaGxBreast/EORTC10994.rda",     "MetaGxBreast/EXPO.rda",
    "MetaGxBreast/FNCLCC.rda",     "MetaGxBreast/GSE25066.rda",     "MetaGxBreast/GSE32646.rda",
    "MetaGxBreast/GSE48091.rda",     "MetaGxBreast/GSE58644.rda",     "MetaGxBreast/HLP.rda",
    "MetaGxBreast/IRB.rda",           "MetaGxBreast/KOO.rda",     "MetaGxBreast/LUND.rda",
    "MetaGxBreast/LUND2.rda",     "MetaGxBreast/MAINZ.rda",     "MetaGxBreast/MAQC2.rda",
    "MetaGxBreast/MCCC.rda",     "MetaGxBreast/MDA4.rda", "MetaGxBreast/METABRIC.rda",
    "MetaGxBreast/MSK.rda", "MetaGxBreast/MUG.rda", "MetaGxBreast/NCCS.rda", 
    "MetaGxBreast/NCI.rda", "MetaGxBreast/NKI.rda", "MetaGxBreast/PNC.rda",
    "MetaGxBreast/STK.rda", "MetaGxBreast/STNO2.rda", "MetaGxBreast/TCGA.rda",
    "MetaGxBreast/TRANSBIG.rda", "MetaGxBreast/UCSF.rda", "MetaGxBreast/UNC4.rda",
    "MetaGxBreast/UNT.rda", "MetaGxBreast/UPP.rda", "MetaGxBreast/VDX.rda"
  ),
  BiocVersion="3.7",
  Genome=c("Affymetrix HGU", "Affymetrix HGU GPL570", "Affymetrix HGU GPL570", "Affymetrix HGU GPL570","Affymetrix HGU95 GPL8300","Affymetrix X3P GPL1352",
           "Affymetrix HGU GPL570","Affymetrix HGU GPL96","Affymetrix HGU GPL570","In-house cDNA GPL4819", "Affymetrix HGU GPL96","Affymetrix HGU GPL570","Affymetrix RSTA GPL10379","Affymetrix Gene1.0ST GPL6244",
           "Illumina","Affymetrix HGU GPL570","Affymetrix HGU95","Swegene GPL14373","Swegene GPL3883","Affymetrix HGU GPL96","Affymetrix HGU GPL96///GPL570///GPL1261",
           "Illumina GPL6106","Affymetrix HGU","Affymetrix HGU GPL96", "Operon GPL6486", "Affymetrix HGU GPL96", "In-house cDNA",
           "Agilent", "Affymetrix HGU GPL570", "Affymetrix HGU GPL97///GPL96", "In-house cDNA GPL180///GPL2776///GPL2777",
           "TCGA", "Affymetrix HGU GPL96", "In-house cDNA", "Agilent GPL885///GPL887///GPL1390", "Affymetrix HGU",
           "Affymetrix HGU GPL570", "Affymetrix HGU", "METABRIC"),
  SourceType="RData",
  SourceUrl=c("http://www.ebi.ac.uk/arrayexpress/experiments/E-TABM-158/", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE19615",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE18864", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE3744",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse3143", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse6861",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse12276", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE1561",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE2109", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE7017",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE25066", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE32646",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE48091", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE58644",
              "http://www.ebi.ac.uk/arrayexpress/experiments/E-TABM-543/", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE5460",
              "Unavailable", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE31863",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE5325", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE11121",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE20194","http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE19177",
              "http://bioinformatics.mdanderson.org/pubdata.html", "https://www.ebi.ac.uk/ega/studies/EGAS00000000083",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse2603", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse10510",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse5364", "Supplemental data from paper Breast cancer classification and prognosis based on gene expression profiles from a population-based study",
              "Not available", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE20711", 
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse1456", "http://smd.princeton.edu/cgi-bin/publication/viewPublication.pl?pub_no=248",
              "http://cancergenome.nih.gov/", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gsE7390",
              "Not available", "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse18229",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse2990","http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse3494",
              "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse2034; http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse5327"),
  SourceVersion="Feb 02 2018",
  Species="Homo sapien",
  TaxonomyId=9606,
  Coordinate_1_based=FALSE,
  DataProvider=c("Array Express", rep("GEO", 13), "Array Express", "GEO", "NA", rep("GEO", 5), "MDAnderson", "EGA", 
                 rep("GEO", 3), "paper supplementary", "NA", "GEO", "GEO", "SMD", "TCGA", "GEO", "NA", rep("GEO", 4)),
  Maintainer="Michael Zon <michaelzon7@gmail.com>",
  RDataClass="ExpressionSet",
  DispatchClass="Rda",
  Tags = "breast cancer expression",
  ResourceName = c(paste0(infoString, ".rda"))
)

write.csv(file="metadata.csv", main.data, row.names=FALSE)



R/loadBreastEsets.R
#' Function to load breast cancer expression sets from the Experiment Hub
#'
#' This function returns breast cancer datasets from the hub and a vector of patients from the datasets that are most likely duplicates
#' @param loadString a character vector specifying which data will be loaded. The default is "majority", which loads in 37 of the 39 datasets.
#' The other option is to provide a character vecotr of the names of the datasets to load. The metabric and tcga datasets areloaded separately as they are 
#' very large and doing so will help prevent memory allocation errors for R windows. Furthermore, these datasets are so large that they
#' dominate statistical analyses so it is best that they are analyzed separate of the 37 smaller datasets loaded with the string majority
#' @param removeDuplicates remove patients with a Spearman correlation greater than or equal to 0.98 with other patient expression profiles (default TRUE)
#' @param quantileCutoff A nueric between 0 and 1 specifying to remove genes with standard deviation below the required quantile (default 0)
#' @param rescale apply centering and scaling to the expression sets (default FALSE)
#' @param minNumberGenes an integer specifying to remove expression sets with less genes than this number (default 0)
#' @param minNumberEvents an integer specifying how man survival events must be in the dataset to keep the dataset (default 0)
#' @param minSampleSize an integer specifying the minimum number of patients required in an eset (default 0)
#' @param removeRetracted remove datasets from retracted papers (default TRUE, currently just PMID17290060 dataset)
#' @param removeSubsets remove datasets that are a subset of other datasets (defeault TRUE, currently just PMID19318476)
#' @param keepCommonOnly remove probes not common to all datasets (default FALSE)
#' @param imputeMissing remove patients from datasets with missing expression values
#' @return a list with 2 elements. The First element named esets contains the datasets. The second element named duplicates contains
#' a vector with patient IDs for the duplicate patients (those with  Spearman correlation greater than or equal to 0.98 with other patient expression profiles).
#' @export
#' @importFrom Biobase esApply featureNames sampleNames exprs pData experimentData
#' @importFrom lattice levelplot
#' @importFrom impute impute.knn
#' @importFrom ExperimentHub ExperimentHub
#' @importFrom AnnotationHub query
#' @importFrom stats complete.cases sd quantile
#' @examples
#'
#' #Use the default loadString = "majority" if you want the 37 smaller datasets
#' esetsAndDups = loadBreastEsets(loadString = c("CAL", "DFHCC", "DFHCC2", "DFHCC3", "DUKE", "DUKE2", "EMC2"))


loadBreastEsets = function(loadString = "majority", removeDuplicates = TRUE, quantileCutoff = 0, rescale = FALSE, minNumberGenes = 0,
                           minNumberEvents = 0, minSampleSize = 0, removeRetracted = TRUE, removeSubsets = TRUE,
                           keepCommonOnly = FALSE, imputeMissing = FALSE)
{
  duplicates = NULL
  #if(getRversion() >= "2.15.1")  utils::globalVariables(c("."), add = F)
  ## -----------------------------------------------------------------------------
  ## needed functions
  ## -----------------------------------------------------------------------------
  filterQuantile <- function(object, q){
    if (!identical(q >=0 && q < 1, TRUE))
      stop("require 0 <= q < 1")
    if (!identical(class(object) == "ExpressionSet", TRUE))
      stop("object must be an ExpressionSet")
    geneSd <- Biobase::esApply(object,1,sd, na.rm=TRUE)
    gene.quantile <- stats::quantile(geneSd, probs=q)
    actual.makescutoff <- sum(geneSd < gene.quantile) / length(geneSd)
    ##make sure the correct number of genes are getting filtered:
    if (abs(q - actual.makescutoff) > 0.01){
      stop("Not scaling this object, likely pre-scaled.")
    }else{
      object <- object[geneSd > gene.quantile, ]
    }
    return(object)
  }
  ##recursive intersect function
  intersectMany <- function(lst){
    ## Find the intersection of multiple vectors stored as elements of a
    ## list, through a tail-recursive function.
    if (length(lst)==2){
      return(intersect(lst[[1]],lst[[2]]))
    }else{
      return(intersectMany(c(list(intersect(lst[[1]],lst[[2]])),lst[seq(-1, -2)])))
    }
  }
  
  ##Split out non-specific probe sets
  expandProbesets <- function (eset, sep = "///"){
    x <- lapply(Biobase::featureNames(eset), function(x) strsplit(x, sep)[[1]])
    eset <- eset[order(vapply(x, length, numeric(1))), ]
    x <- lapply(Biobase::featureNames(eset), function(x) strsplit(x, sep)[[1]])
    idx <- unlist(vapply(x, function(i) rep(i, length(x)), character(length(x))))
    xx <- !duplicated(unlist(x))
    idx <- idx[xx]
    x <- unlist(x)[xx]
    eset <- eset[idx, ]
    Biobase::featureNames(eset) <- x
    eset
  }
  
  ## -----------------------------------------------------------------------------
  ##load the esets
  ## -----------------------------------------------------------------------------
  
  hub = ExperimentHub::ExperimentHub()
  #AnnotationHub::possibleDates(hub)
  #ovarianData = query(hub, c("MetaGxOvarian", "ExpressionSet"))
  breastData = query(hub, c("MetaGxBreast", "ExpressionSet"))
  tcgaInd = which(grepl("TCGA", breastData$title))
  metabricInd = which(grepl("METABRIC", breastData$title))
  if(length(loadString) == 1)
  {
    if(loadString == "majority"){
      breastData = breastData[seq_len(length(breastData))[-c(metabricInd, tcgaInd)]]
    }else{
      stop("loadString needs to be one of majority, metabric, tcga, or a character vector of datasets to load from the hub")
    }
  }else{
    keepIndVec = c()
    for(i in seq_len(length(loadString)))
    {
      keepInd = which(grepl(paste0(loadString[i], "_"), paste0(breastData$title, "_")))
      if(length(keepInd) == 0){
        stop(paste(loadString[i], "could not be found in the MetaGxBreast package in the experiment hub"))
      }else{
        keepIndVec = c(keepIndVec, keepInd) 
      }
    }
    breastData = breastData[keepIndVec]
  }

  
  
  esets <- list()
  #if()
  for(i in seq_len(length(breastData)))
  {
    dataName = breastData[i]$title
    esets[[i]] = breastData[[names(breastData)[i]]]
    names(esets)[i] = breastData[i]$title
  }
  
  ## -----------------------------------------------------------------------------
  ##Explicit removal of samples from specified datasets:
  ## -----------------------------------------------------------------------------
  delim <- ":"   ##This is the delimiter used to specify dataset:sample,
  
  ## same as used in metagx getbrcadata
  #load("inst\\extdata\\BenDuplicate.rda")
  #source(system.file("extdata", "patientselection.config", package="MetaGxBreast"))
  load(system.file("extdata", "duplicates.rda", package="MetaGxBreast"))
  
  rmix <- duplicates
  ii <- 1
  while (length(rmix) > ii){
    rmix <- rmix [!is.element(names(rmix), rmix[[ii]])]
    ii <- ii+1
  }
  rmix <- unique(unlist(rmix))
  
  message("Clean up the esets.")
    
  for (i in seq_len(length(esets))){
    eset <- esets[[i]]
    
    ##filter genes with standard deviation below the required quantile
    if(quantileCutoff > 0 && quantileCutoff < 1){
      eset <- filterQuantile(eset, q=quantileCutoff)
    }
    ##rescale to z-scores
    if(rescale == TRUE){
      Biobase::exprs(eset) <- t(scale(t(Biobase::exprs(eset))))
    }
    
    if(removeDuplicates == TRUE){
      keepix <- setdiff(colnames(eset@assayData$exprs), rmix)
      if(length(keepix) != length(colnames(eset@assayData$exprs)))
      {
        newEset = ExpressionSet(Biobase::exprs(eset)[, keepix, drop=FALSE])
        newEset@experimentData = eset@experimentData
        newEset@phenoData = eset@phenoData
        newEset@phenoData@data = Biobase::pData(eset)[keepix, , drop=FALSE]
        newEset@featureData = eset@featureData
        eset = newEset
      }
      #Biobase::exprs(eset) <- Biobase::exprs(eset)[, keepix, drop=FALSE]
      #Biobase::pData(eset) <- Biobase::pData(eset)[keepix, , drop=FALSE]
      
    }
    
    ##include study if it has enough samples and events:
    if (!is.na(minNumberEvents)
        && exists("minSampleSize") && !is.na(minSampleSize)
        && minNumberEvents > 0
        && sum(eset$vital_status == "deceased") < minNumberEvents
        || ncol(eset) < minSampleSize)
    {
      message(paste("excluding",
                    "(minNumberEvents or minSampleSize)"))
      next
    }
    if(nrow(eset) < minNumberGenes) {
      message(paste("excluding experiment hub dataset",breastData[i]$title,"(minNumberGenes)"))
      next
    }
    if(removeRetracted && length(grep("retracted", Biobase::experimentData(eset)@other$warnings$warnings)) > 0){
      message(paste("excluding experiment hub dataset",breastData[i]$title,"(removeRetracted)"))
      next
    }
    if(removeSubsets && length(grep("subset", Biobase::experimentData(eset)@other$warnings$warnings)) > 0){
      message(paste("excluding experiment hub dataset",breastData[i]$title,"(removeSubsets)"))
      next
    }
    message(paste("including experiment hub dataset",breastData[i]$title))
    ##    featureNames(eset) <- make.names(featureNames(eset))  ##should not do this, it is irreversible.
    esets[[i]] <- eset
    rm(eset)
  }
    
  
  ##optionally take the intersection of genes common to all platforms:
  if(keepCommonOnly){
    features.per.dataset <- lapply(esets, Biobase::featureNames)
    intersect.genes <- intersectMany(features.per.dataset)
    esets <- lapply(esets, function(eset){
      eset <- eset[intersect.genes, ]
      return(eset)
    })
  }
  
  ids.with.missing.data <- which(vapply(esets, function(X)
    sum(!complete.cases(Biobase::exprs(X))) > 0, numeric(1)) == 1)
  message(paste("Ids with missing data:", paste(names(ids.with.missing.data),
                                                collapse=", ")))
  
  if (length(ids.with.missing.data) > 0 && imputeMissing) {
    for (i in ids.with.missing.data) {
      Biobase::exprs(esets[[i]]) = impute::impute.knn(Biobase::exprs(esets[[i]]))$data
    }
  }
  
  retList = list(esets, duplicates)
  names(retList) = c("esets", "duplicates")
  return(retList)
}
